src/math/Math.sol

```diff
@@ -32,4 +32,40 @@ library Math {
             z := add(gt(mod(x, y), 0), div(x, y))
         }
     }
+
+    /// @dev Returns the floor of log2(x) and returns 0 on input 0.
+    /// @dev This computation makes use of De Bruijn sequences, their usage for computing log2 is referenced here: http://supertech.csail.mit.edu/papers/debruijn.pdf
+    function log2(uint256 x) internal pure returns (uint256 y) {
+        assembly {
+            // Use the highest set bit of x to set each of its lower bits
+            x := or(x, shr(1, x))
+            x := or(x, shr(2, x))
+            x := or(x, shr(4, x))
+            x := or(x, shr(8, x))
+            x := or(x, shr(16, x))
+            x := or(x, shr(32, x))
+            x := or(x, shr(64, x))
+            x := or(x, shr(128, x))
+            // Take only the highest bit of x
+            x := xor(x, shr(1, x))
```

Comment:
The original idea comes from here: https://graphics.stanford.edu/~seander/bithacks.html#IntegerLogDeBruijn

Less control flow is generally a good idea for efficiency. We've seen it for the other math operations where we removed the `if` statements. I think it computes exactly the same thing otherwise so I don't think it uses another assumption like not needing to set all the bits to 0

Reply:
Indeed, it works much better.